### PR TITLE
Share Image part #2 (W/ Bootstrap Modal)

### LINKED ISSUE #11: Share Image
#### As a user I want to share an image with someone via email.

__Story__:

This application is called Image Sharer. Sure, you can share images by adding
them to our web service, but how will people who do not know about this web
service be able to see those images? In phone calls with our customers we have
discovered that many of our users are manually emailing their friends links to
the images. That is awesome! We can make things more efficent for our users by
directly supporting sharing images through email.

__Acceptance criteria__:
- [x] There is a link to share an image.
- [x] After clicking on this link, a bootstrap modal appears with a form to
  send the email.
- [x] The form has fields for the recipient's email and an optional message.
- [x] The email we send has an HTML part and a text part.
- [x] The email contains the image, the custom message if provided, and a link
  inviting the recipient to check out our application.
- [x] The modal closes and a success message is shown if the email was sent
  successfully.
- [x] The user is alerted of errors if the operation was unsuccessful.

__Note__: This story is large, so it makes sense to break the implementation
down into more manageable pieces (a la elephant carpaccio). Please start by
implementing a basic sharing feature that uses a new page (not a modal), and
open a PR for that. Then, once we've verified the simplified version is
working, refactor it to use a modal, and open a second PR for that. The
[video](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project/basics-of-bootstrap-modals-js-and-ajax)
provides some background and code examples that will be really helpful for the
second PR, so it's worth watching (and following along) before you begin that
part.

__Discussion topics__:
- [x] How could we develop the front end and back end separately? (And why
  might that be a good idea?)
- [x] How should we implement a non-CRUD action?
- [x] What are the different types of errors we should handle?
- [x] What is AJAX? What libraries can help us add AJAX to our Rails app?
- [x] How should we go about sending emails from our Rails app?
- [x] How could we show a preview of the image being shared in the modal?
- [x] What are the implications of allowing users to send emails from our app?

__Reference material__:
- http://guides.rubyonrails.org/action_mailer_basics.html
- https://elements.heroku.com/addons/sendgrid
- https://sendgrid.com/docs/Integrate/Frameworks/rubyonrails.html
- https://github.com/rails/jquery-ujs/wiki
- http://v4-alpha.getbootstrap.com/components/modal
- https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project/basics-of-bootstrap-modals-js-and-ajax

__Dependencies__:
- Image Index
- Refactor flow tests to use page objects

### PR DESCRIPTION
**Pull request by almgong**
Closes #11

### CODE REVIEW

**Comment by lsq921224** [comment]
Overall looks great, there is one more edge case though. If we try to share a nonexistent image, the action failed and the share button is greyed out. But there is no alert, which is against the acceptance criteria: `The user is alerted of errors if the operation was unsuccessful.`, can we handle this case and write a flow test for it?
You can choose how exactly you want to handle this case, but I would at least want to see an error message as a user.

File: app/controllers/share_controller.rb [hunk 1]
```diff
@@ -20,9 +20,16 @@ def share_params
...
-    message = email_valid ? 'Email successfully sent.' : 'Invalid email address!'
-    status = email_valid ? :ok : :unprocessable_entity
+  def get_share_response(options)
+    share_valid = options[:to_address].to_s.present? && options[:image].present?
+
+    message = 'Email successfully sent.'
+    status = :ok
+
+    unless share_valid
+      message = options[:image].present? ? 'Invalid email address!' : 'Invalid image, please try again.'
```
Comment (lsq921224) [inline app/controllers/share_controller.rb]: if the image is not present, no matter how many times our users retry, they will not be able to share this image.. 
Even though we are not UX people, but when we work on stories, we want to wear our customer's hat, so we can better deliver what our customers will like.. However I did say you can choose how exactly you want to handle this case as long as you give me a flash message, so fine..
As a side note, in APM, if something like this happens, we will usually redirect users to the corresponding index page with a flash message

File: test/flows/share_flow_test.rb [hunk 2]
```diff
@@ -50,4 +50,38 @@ class ShareFlowTest < FlowTestCase
...
+    images_index_page = PageObjects::Images::IndexPage.visit
+
+    image_to_share = images_index_page.images.find do |image|
+      image.url == mind_blown_gif
+    end
+
+    image_to_share.click_share
+
+    # open a new window and delete the image before share completes
+    within_window open_new_window do
```
Comment (lsq921224) [inline test/flows/share_flow_test.rb]: This is interesting, I've never seen this before, but it looks like it does what it's supposed to do.. I like how you try to find a way to actually open a new tab and delete the image, normally people simply destroy the created image.
